jax/experimental/optimizers.py

```diff
@@ -229,6 +229,44 @@ def get_params(state):
     return x
   return init, update, get_params
 
+
+@optimizer
+def adagrad(step_size, momentum=0.9):
+  """Construct optimizer triple for Adagrad.
+
+  Adaptive Subgradient Methods for Online Learning and Stochastic Optimization:
+  http://www.jmlr.org/papers/volume12/duchi11a/duchi11a.pdf
+
+  Args:
+    step_size: positive scalar, or a callable representing a step size schedule
+      that maps the iteration index to positive scalar.
+    momentum: optional, a positive scalar value for momentum
+
+  Returns:
+    An (init_fun, update_fun, get_params) triple.
+  """
+  step_size = make_schedule(step_size)
+
+  def init(x0):
+    g_sq = np.zeros_like(x0)
+    m = np.zeros_like(x0)
+    return x0, g_sq, m
+
+  def update(i, g, state):
+    x, g_sq, m = state
+    g_sq += g**2
+    g_sq_inv_sqrt = np.where(g_sq > 0, 1. / np.sqrt(g_sq), 0.0)
```

Comment:
potentially could improve runtime by forcing in place calculation?

Reply:
With XLA there's no way to force in-place calculation, but it will happen automatically under a `jit`, so things are fast by default!